From cee4a40e7febe8f1b5b38b31f42f1b64f536ddb4 Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Tue, 4 Apr 2017 16:21:45 +0200
Subject: [PATCH] tc: m_xt: Fix segfault when adding multiple actions at once

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1326726
Upstream Status: iproute2.git commit 8eee75a8358c5

commit 8eee75a8358c542d881d2a84e2c47cc0a9fa92ef
Author: Phil Sutter <phil@nwl.cc>
Date:   Fri Jun 10 13:42:01 2016 +0200

    tc: m_xt: Fix segfault when adding multiple actions at once

    Without this, the following call to tc would segfault:

    | tc filter add dev d0 parent ffff: u32 match u32 0 0 \
    |       action xt -j MARK --set-mark 0x1 \
    |       action xt -j MARK --set-mark 0x1

    The reason is basically the same as for 6e2e5ec28bad4 ("fix print_ipt:
    segfault if more then one filter with action -j MARK.") but in
    parse_ipt() instead of print_ipt().

    Signed-off-by: Phil Sutter <phil@nwl.cc>
---
 tc/m_xt.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/tc/m_xt.c b/tc/m_xt.c
index aa83061..f0efd28 100644
--- a/tc/m_xt.c
+++ b/tc/m_xt.c
@@ -133,7 +133,9 @@ static int parse_ipt(struct action_util *a,int *argc_p,
 	__u32 hook = 0, index = 0;
 	struct option *opts = NULL;
 
-	xtables_init_all(&tcipt_globals, NFPROTO_IPV4);
+	/* copy tcipt_globals because .opts will be modified by iptables */
+	struct xtables_globals tmp_tcipt_globals = tcipt_globals;
+	xtables_init_all(&tmp_tcipt_globals, NFPROTO_IPV4);
 	set_lib_dir();
 
 	{
@@ -152,7 +154,7 @@ static int parse_ipt(struct action_util *a,int *argc_p,
 	}
 
 	while (1) {
-		c = getopt_long(argc, argv, "j:", tcipt_globals.opts, NULL);
+		c = getopt_long(argc, argv, "j:", tmp_tcipt_globals.opts, NULL);
 		if (c == -1)
 			break;
 		switch (c) {
@@ -165,12 +167,12 @@ static int parse_ipt(struct action_util *a,int *argc_p,
 					return -1;
 				}
 #if (XTABLES_VERSION_CODE >= 6)
-			opts = xtables_options_xfrm(tcipt_globals.orig_opts,
-						    tcipt_globals.opts,
+			opts = xtables_options_xfrm(tmp_tcipt_globals.orig_opts,
+						    tmp_tcipt_globals.opts,
 						    m->x6_options,
 						    &m->option_offset);
 #else
-			opts = xtables_merge_options(tcipt_globals.opts,
+			opts = xtables_merge_options(tmp_tcipt_globals.opts,
 						     m->extra_opts,
 						     &m->option_offset);
 #endif
@@ -178,7 +180,7 @@ static int parse_ipt(struct action_util *a,int *argc_p,
 				fprintf(stderr, " failed to find additional options for target %s\n\n", optarg);
 				return -1;
 			} else
-				tcipt_globals.opts = opts;
+				tmp_tcipt_globals.opts = opts;
 			} else {
 				fprintf(stderr," failed to find target %s\n\n", optarg);
 				return -1;
-- 
1.8.3.1

